text/0249-v2-experiments.md

```diff
@@ -199,6 +199,82 @@ Disadvantages:
   mitigate this downside by keeping the old unstable package around, but that
   leads to duplicated classes.
 
+### Option 5: superset/subset releases
+
+Instead of *splitting* the stable and experimental APIs into two different
+packages, we release two builds of the same source code, where the *experimental*
+build is a superset of the *stable* build.
+
+This can be done automatically, by using jsii to strip out experimental APIs
+from the **public interface** of the stable release. The implementation
+(`.js` files) will be the same in both releases, so stable APIs in
+`aws-cdk-lib` can safely use experimental APIs internally (though not in
+their public API surfaces).
+
+**How does stripping APIs from the public work?**
+
+| TypeScript            | JavaScript | Java, Python, C#, Go            |
+|-----------------------|------------|---------------------------------|
+| Stripped from `.d.ts` | N/A        | No bindings generated by `jsii` |
+
+In TypeScript, the compiler will prevent usage of APIs that are not advertised
+in the `.d.ts` files. Users will be able to cast this away using `as any` though.
+They will also be able to confuse the compiler enough that they will get to pass
+struct properties that are technically `@experimental`.
+
+For JavaScript, an IDE will (hopefully) not autocomplete APIs missing from
+the `.d.ts` files, but otherwise nothing prevents users from calling the
+APIs, or passing struct properties that are technically experimental. On the
+other hand, JavaScript devs are by definition required to read the docs,
+so they can't miss the *experimental* banners displayed there.
+
+For jsii client languages, the bindings for experimental APIs will simply not
+exist so there's no way to work around that (though we'll have to check
+whether for example Python makes it possible to pass struct values that
+aren't in the declared API surface).
+
+**Vending**
+
+There are a number of different ways we can choose to distinguish the two builds.
+See Appendix D for an evaluation of all the possibilities.
+
+- Distinguish by package name
+  - Can't be used in **pip**: there's no way to have an app that uses the experimental
+    library work with a library that uses the stable library (no way to prevent the
+    transitive dependency package with a different name from being installed).
```

Comment:
I don't understand exactly why that is... I thought that if a JSII-enabled JS/TS library had a peer dependency on `aws-cdk-lib`, then its `pip` version would **not** have a dependency on the Python version of `aws-cdk-lib`, because the customer of the app is supposed to bring in a dependency on the Python version of `aws-cdk-lib`, in whatever version they want? Is my understanding incorrect, and a peer-only dependency in JavaScript gets translated in Python to a regular dependency?

Or is the problem here a Python-only construct library?

Reply:
To my understanding, we have `peerDependencies` only because NPM allows multiple copies, which we don't want. A `peerDependency` still allows to model a dependency with an explicit version number though.

In other package managers, we don't have the problem of multiple copies so we don't need the solution of `peerDependencies`. In fact it doesn't exist, and the only way to achieve the same effect in `pip` would be NOT declare a dependency at all. Then you still require the consumer to bring the dependency (good but unnecessary), but the package cannot declare what version it needs (bad).

-------------

All this is a long way to say: `peerDependencies` in npm will be translated into *regular* dependencies in other package managers.

Reply:
Does this imply that the solution you gave in the Maven part (`provided` scope) is not viable then?